Validate bundle directive (#53)

rnpridgeon opened the pull request:
Compliments: https://github.com/suborbital/atmo/pull/137

Subo does not currently verify the directive when bundling projects which may result in unexpected errors when deploying a bundle.

This PR adds an additional step to the build process which will validate the augmented directive. 

User defined directive
```
handlers:
  - type: request
    resource: /hello
    method: POST
    steps:
      - fn: helloworld
  - type: request
    resource: /hello
    method: POST
    steps:
      - fn: helloworld
```

Without this change:
```
subo build .
ℹ️  🐳 using Docker toolchain
⏩ START: building runnable: helloworld (rust)
 Downloading crates ...
...
✅ DONE: helloworld was built -> /root/runnable/helloworld/helloworld.wasm
✅ DONE: bundle was created -> /home/ryanp/Workspaces/suborbital/subo/demo/runnables.wasm.zip @ v0.1.0
```

With this change:
```
subo build .
ℹ️  🐳 using Docker toolchain
⏩ START: building runnable: helloworld (rust)
 Downloading crates ...
 ...
✅ DONE: helloworld was built -> /root/runnable/helloworld/helloworld.wasm
Error: failed to Bundle: 🚫 failed to AugmentAndValidateDirectiveFns: found 1 problems:
        a handle is already registered for path '/hello'
Usage:
  subo build [dir] [flags]

```

rnpridgeon commented:
Turns out this is a bit more aggressive than it should be. Will revisit monday

rnpridgeon commented on builder/context/buildcontext.go:
```diff
@@ -73,7 +73,9 @@ func ForDirectory(dir string) (*BuildContext, error) {
 
 	directive, err := readDirectiveFile(fullDir)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to readDirectiveFile")
```
note: 

I moved this out of the `readDirectiveFile` method to the build context because it felt out of place. 

IMO if the caller has expressed the intent to read the directive but it can't be found this is an error; it is the caller's (build context) responsibility to decide what they want to do in this event.

rnpridgeon commented on scripts/sanity_project.sh:
```diff
@@ -0,0 +1,30 @@
+  if [[ -d "$TEST_PROJECT" ]]; then
+    echo "cleaning up test artifacts"
+    # target directory created in docker has root owner
+    sudo rm -rf "$TEST_PROJECT"
```
Note: This is sort of annoying but the docker build context ends up doing everything as root which means the filesystem objects are ultimately owned by root. 

This is worth addressing in another PR.

jagger27 replied:
I'm not seeing this issue on macOS (probably due to how Docker volumes just straight up ignore permissions on macOS).

rnpridgeon replied:
fair enough, I'll back it out and leave a comment about how this may run into permission errors on cleanup

rnpridgeon commented on builder/context/directive.go:
```diff
@@ -35,10 +35,6 @@ func readDirectiveFile(cwd string) (*directive.Directive, error) {
 
 	_, err := os.Stat(filePath)
 	if err != nil {
-		if os.IsNotExist(err) {
```
note:

I moved this out of the readDirectiveFile method to the build context because it felt out of place.

IMO if the caller has expressed the intent to read the directive but it can't be found this is an error; it is the caller's (build context) responsibility to decide what they want to do in this event.

javorszky commented:
The very first piece of review of whether we should find a different expression for "sanity". Reason: https://www.selfdefined.app/definitions/sanity-check/ (tldr it's ableist)

javorszky commented on scripts/sanity_runnable.sh:
```diff
@@ -0,0 +1,34 @@
+  fi
+
+  # return to origin, clear directory stack
+  pushd -0 > /dev/null && dirs -c
```
TIL about `pushd` and `popd`! https://opensource.com/article/19/8/navigating-bash-shell-pushd-popd

Thank you! 🙂

rnpridgeon replied:
one of the most useful bash builtins

javorszky commented on scripts/sanity_runnable.sh:
```diff
@@ -0,0 +1,34 @@
+subo create runnable swift-test --lang swift
+subo create runnable as-test --lang assemblyscript
+subo create runnable tinygo-test --lang tinygo
+subo create runnable js-test --lang javascript
```
if my understanding is correct, if this completes successfully, the files would be left in place, rather than cleaned up after.

Is that expected / desired?

rnpridgeon replied:
The trap is triggered on exit (error or otherwise) which cleans up the root project directory if present.

set -e tells the script to terminate on error, set -u tells the script to error if a variable isn't bound to a value (useful when you have potentially dangerous commands like `rm -rf $var`

rnpridgeon replied:
I do need to comment out javascript for  now since it will fail to bundle until the new images are released

javorszky replied:
Had a conversation on discord, the `trap` takes care of running the `catch` function to clean up whatever happens (error or success)

rnpridgeon commented on Makefile:
```diff
@@ -15,6 +15,14 @@ subo/docker:
 subo/docker/publish:
 	docker buildx build . --platform linux/amd64,linux/arm64 -t suborbital/subo:dev --push
 
+subo/sanity: subo/sanity/project subo/sanity/runnable
```
yep,  I had not considered this before but it makes sense. I'm not sure if we can change `sanity.yml` from which I borrowed this term but I have update the scripts and target names accordingly.

rnpridgeon commented:
renamed targets and scripts to `smoketest`, removed sudo from clean up trap. 

I'm not sure if `sanity.yml` is a required name for the CI so I'll leave that as is for now and we can follow up on it later. 

https://github.com/suborbital/subo/issues/212

javorszky reviewed (approved):
Thanks for fixing these! 🙂 

:shipit:
